Implement an audio HAL based custom board with support for both PCM51xx and MA210x0

douardda opened the pull request:
This is an attempt at abstracting the handling of the DAC partially using the Audio HAP API os the ESP-ADF toolkit.

The PCM51xx seems to be ok and is mostly working, but the MA210x0 has not been tested, and will probably need some more care to get back to its working condition before this PR.

@jorgenkraghjakobsen please tell me what you think of this PR, if it seems going in the right direction.

Note: it depends in #17 


This defines a custom HAL board that can use a configurable DAC chipset
(PCM51XX and Infineon Merus Audio MA120X0 for now).

To use a custom board, one must:

- select "Audio HAL" -> "Audio board" -> "Custom Board", then
- choose the dac in "Custom Audio Board" -> "DAC chip"
- select I2C pins for the DAC "Custom Audio Board" -> "DAC I2C control
  interface"
- select I2S pins in "Custom Audio Board" -> "I2S master interface"

In this revision:

- component/MerusAudio has been integrated in components/custom_board as
  components/custom_board/ma210x0,
- the Merus (MA120x0) driver now partially implements the audio HAL API
  (i2c configuration and defines an audio_hal struct
  AUDIO_CODEC_MA120X0_DEFAULT_HANDLE),
- dsp_processor now depends on the audio_board component and uses a bit
  of the audio HAL API to configure the i2s interface.
- all MerusAudio specific code has been removed from main.c in favor of
  the audio HAL API.

jorgenkraghjakobsen commented:
Hi David  
Also think that is the correct way to handle this. My org intention was to stay a way from the ADF stuff and keep IDF comp. can not see if that has already been broken. On the other hand I have a MA120x0 HAL driver on my do list anyway and I am already on our gen2 MA120 so that is needed any how. To late now. - I will look thrugh you code tomorrow. Regards Jørgen

douardda commented:
@jorgenkraghjakobsen this does not use the ADF API other than audio HAL API to describe the codec driver.

I've rebased the PR on master  (and fixed a couple of typos in a Kconfig file).

About the ma120 code, I've indeed seen 2 versions of some code (ma120 and ma120x0) and wasn't sure which to use/keep, so maybe I made wrong stuff there. 

Hopefully this can help you anyway :-)

douardda commented on components/custom_board/ma120x0/MerusAudio.c:
```diff
@@ -51,48 +49,75 @@ static const char* I2C_TAG = "i2c";
+    uint8_t rxbuf[12];
+    ma_read(0x20, 2, 0x0003, rxbuf, 3);
+    // printf("\nVolume : 0x%02x 0x%02x 0x%02x\n", rxbuf[0], rxbuf[1], rxbuf[2]);
+	*vol = rxbuf[0];  // XXX which register? also 128+/-?
```
Note the XXX here, not sure the correct implementation for this function.

jorgenkraghjakobsen replied:
Need conversion here - Read out from 2 volume register is needed or must be synced. Adr 0x0003 ch0 0x0004 ch1. 
0x18 is 0 dB gain on volume and and each step up 0.75 dB attenuation - Amplifier has 20 dB build in. 
Just let me take over from here.

douardda commented on components/custom_board/ma120x0/MerusAudio.c:
```diff
@@ -51,48 +49,75 @@ static const char* I2C_TAG = "i2c";
+esp_err_t ma120x0_set_mute(bool enable)
+{
+	esp_err_t ret = ESP_OK;
+	// TODO
```
Same, I have not searched how to implement these 2 functions.

douardda commented on main/main.c:
```diff
@@ -512,35 +505,17 @@ void app_main(void)
-    ma_read(0x20,2,0x0003, rxbuf,3);
-    printf("\nVolume : 0x%02x 0x%02x 0x%02x\n",rxbuf[0], rxbuf[1], rxbuf[2] );
-    ma_write_byte(0x20,2,0x608, 51);
+    // ma_write_byte(0x20,2,0x608, 51); // XXX what's this for?
```
Note the XXX also here, no idea why/what this is for.

jorgenkraghjakobsen replied:
Early prototype hack - internal RC oscillator trimming. Not needed anymore.

douardda commented on components/dsp_processor/dsp_processor.c:
```diff
@@ -57,42 +58,34 @@ void setup_dsp_i2s(uint32_t sample_rate, bool slave_i2s)
-    .data_out_num =  CONFIG_SLAVE_I2S_DATAOUT_PIN,
-    .data_in_num = -1                                                       //Not used
-  };
+  /* is this needed?
```
I've commented this section to prevent using directly CONFIG_MASTER_I2S_XXX variables here, not sure if these statements are still needed in this version of the code.

jorgenkraghjakobsen replied:
Still needed in an non ADF environment - Think i need to do a menuconfig to enable ADF or IDF compatibility

miLORD1337 commented on main/Kconfig.projbuild:
```diff
@@ -33,9 +33,9 @@ endmenu
-    config SNAPCLIENT_SNTP_ENABLE
-        string "Enable SNTP"
-        default "true"
+    config SNPACLIENT_SNTP_ENABLE
```
```suggestion
    config SNAPCLIENT_SNTP_ENABLE
```
Typo, I guess.

miLORD1337 commented on components/dsp_processor/component.mk:
```diff
@@ -1,11 +1,11 @@
 
-COMPONENT_SRCDIRS := .
-# CFLAGS +=
+COMPONENT_ADD_INLUCDEDIRS += ./include
```
```suggestion
COMPONENT_ADD_INCLUDEDIRS += ./include
```
Typo.

miLORD1337 commented on components/dsp_processor/Kconfig.projbuild:
```diff
@@ -1,59 +1,22 @@
-        int "Slave i2s data out"
-        default 5
+    config USE_BIQUAD_ASM
+        bool "Use uptimized asm verion of Biquad_f32"
```
```suggestion
        bool "Use optimized asm verion of Biquad_f32"
```
Typo.

miLORD1337 commented on components/dsp_processor/Kconfig.projbuild:
```diff
@@ -1,59 +1,22 @@
-        default 12
-        help
-            Slave audio interface left/right sync clock.
+           Select number of bits per sample for codec configed and connected to esp32 i2s dma hw
```
```suggestion
           Select number of bits per sample for codec configured and connected to esp32 i2s dma hw
```
